notebook/notebook/notebook.py

```diff
@@ -432,7 +436,7 @@ def notebook_status_from_pod(pod):
                 'pod_ip': f'{pod_ip}:{POD_PORT}'
             })
 
-    return web.HTTPNotFound()
+    return web.HTTPForbidden()
```

Comment:
Just so I understand, in aiohttp you can return exceptions or raise them? From what I understand exceptions raised in an awaited coroutine (or future) do not necessarily raise in the async consumer, so it would make sense that aiohttp allows you to return exceptions, but in documentation, I see detail of raising exceptions in request handlers, rather than returning them.

https://docs.aiohttp.org/en/latest/web_exceptions.html#aiohttp.web.HTTPUnauthorized

Reply:
Also, this seems inconsistent with our decorators, which raise rather than return. It may work differently there because it skips invoking the aiohttp request handler.

Reply:
> Just so I understand, in aiohttp you can return exceptions or raise them?

Correct.  My usual practice is to return normal (successful) exit codes if possible and raise error codes.

 > From what I understand exceptions raised in an awaited coroutine (or future) do not necessarily raise in the async consumer

What?  That's not my understanding.  They will be raised (maybe unless it gets cancelled before receiving the raise).  What else could happen?

Reply:
> this seems inconsistent with our decorators, which raise rather than return

The decorators return in the non-exceptional case, for example `return fun(...)` in:

```
def rest_authenticated_users_only(fun):
    async def wrapped(request, *args, **kwargs):
        userdata = await userdata_from_rest_request(request)
        if not userdata:
            raise web.HTTPUnauthorized()
        return await fun(request, userdata, *args, **kwargs)
    return wrapped
```

Reply:
> What?

I think I confused two ways of using asyncio when reading about this last night. We can execute/retrieve the results of a future, task, or coroutine without awaiting on it (using ensure_future, or ensure_task, and attaching a done callback which executes .result()). The exceptions will not be propagated, except under conditions that an exception handler has been set (at least under some kinds of executions). So I was concerned that this may apply to await as well.

https://medium.com/@yeraydiazdiaz/asyncio-coroutine-patterns-errors-and-cancellation-3bb422e961ff

https://www.roguelynn.com/words/asyncio-exception-handling/l: 

" Exceptions will not crash the system - unlike non-asyncio programs. And they might go unnoticed, as we saw when return_exceptions=False, where we failied to save a message causing us to never hit cleanup.

Be sure to set some sort of exception handling – either globally with loop.set_exception_handler, individually like in our handle_results function, or a mix of both (you’ll probably need a mix).
"

but with await, the first exception is immediately propagated to the async task that is awaiting.

Reply:
So yeah, exceptions always propagate through await.

As I understand it, explicit callbacks are not recommended for user-level code.  We don't use them.  E.g. https://docs.python.org/3/library/asyncio-task.html#asyncio.Task.add_done_callback

"This method should only be used in low-level callback-based code."

ensure_future means nobody is waiting for the result of the ... object (coroutine, task, future, future-like, or awaitable?  I don't fully understand the details between all of these.  But at the level we're programming, that is, not low-level, you shouldn't need to.  We should, obviously, although I don't feel like I do yet.)  In that case, if an exception is thrown, it will be printed by the interpreter and the execution of the ... object ... will end.

I think everywhere we ensure_future, the ensured thing catches and handles all exceptions (e..g heal loops, that print the caught exception and retry periodically.)